File: modSrc/common-1.21.8/src/main/java/io/github/summpot/beaconauth/login/ServerLoginNegotiation.java
package io.github.summpot.beaconauth.login;

import net.minecraft.resources.ResourceLocation;

import java.util.HashMap;
import java.util.Map;

/**
 * Tracks BeaconAuth negotiation state for a single Login listener instance.
 */
public final class ServerLoginNegotiation {
    public enum Phase {
        PROBE,
        INIT,
        LOGIN_URL,
        VERIFY,
        COMPLETE
    }

    private final Map<ResourceLocation, LoginQueryType> transactions = new HashMap<>();
    private Phase phase = Phase.PROBE;
    private boolean modded;
    private boolean requiresBeaconAuth;
    private String pendingChallenge;
    private int pendingPort;
    private int ticks;
    private boolean finished;

    public void registerTransaction(ResourceLocation requestKey, LoginQueryType type) {
        transactions.put(requestKey, type);
    }

    public LoginQueryType consume(ResourceLocation requestKey) {
        return transactions.remove(requestKey);
    }

    public void markModded(boolean modded) {
        this.modded = modded;
    }

    public boolean isModded() {
        return modded;
    }

    public void requireBeaconAuth(boolean requires) {
        this.requiresBeaconAuth = requires;
    }

    public boolean requiresBeaconAuth() {
        return requiresBeaconAuth;
    }

    public void setChallenge(String challenge, int port) {
        this.pendingChallenge = challenge;
        this.pendingPort = port;
    }

    public String getPendingChallenge() {
        return pendingChallenge;
    }

    public int getPendingPort() {
        return pendingPort;
    }

    public void setPhase(Phase phase) {
        this.phase = phase;
    }

    public Phase getPhase() {
        return phase;
    }

    public void incrementTick() {
        ticks++;
    }

    public int getTicks() {
        return ticks;
    }

    public void resetTick() {
        ticks = 0;
    }

    public void markFinished() {
        finished = true;
    }

    public boolean isFinished() {
        return finished;
    }
}

File: modSrc/common-1.21.8/src/versioned/kotlin/io/github/summpot/beaconauth/login/ServerLoginNegotiation.kt
package io.github.summpot.beaconauth.login

import net.minecraft.resources.ResourceLocation

/**
 * Tracks BeaconAuth negotiation state for a single Login listener instance.
 *
 * 1.21.x uses cookie request/response where the key is a ResourceLocation.
 */
class ServerLoginNegotiation {
    enum class Phase {
        PROBE,
        INIT,
        LOGIN_URL,
        VERIFY,
        COMPLETE
    }

    private val transactions: MutableMap<ResourceLocation, LoginQueryType> = HashMap()

    var phase: Phase = Phase.PROBE

    private var modded: Boolean = false
    private var requiresBeaconAuth: Boolean = false

    private var pendingChallenge: String? = null
    private var pendingPort: Int = 0

    var ticks: Int = 0
        private set

    private var finished: Boolean = false

    fun registerTransaction(key: ResourceLocation, type: LoginQueryType) {
        transactions[key] = type
    }

    fun consume(key: ResourceLocation): LoginQueryType? = transactions.remove(key)

    fun markModded(modded: Boolean) {
        this.modded = modded
    }

    fun isModded(): Boolean = modded

    fun requireBeaconAuth(requires: Boolean) {
        this.requiresBeaconAuth = requires
    }

    fun requiresBeaconAuth(): Boolean = requiresBeaconAuth

    fun setChallenge(challenge: String, port: Int) {
        this.pendingChallenge = challenge
        this.pendingPort = port
    }

    fun getPendingChallenge(): String? = pendingChallenge

    fun getPendingPort(): Int = pendingPort

    fun incrementTick() {
        ticks++
    }

    fun resetTick() {
        ticks = 0
    }

    fun markFinished() {
        finished = true
    }

    fun isFinished(): Boolean = finished
}
